### DISCUSSION TITLE
I can't import an exported content item.

### DISCUSSION THREAD

**Discussion by aaronamm** [OrchardCMS/OrchardCore#15922] [Q&A] [answered]
Hello everyone.
I have a very simple content type to test if export/import work.
Here is my content type: 
Page content type which contains only TitlePart.
```
Page
- TitlePart
```
- I enabled `Export Content To Deployment Target` features in admin site (Feature menu).
- Then I went back to Page content item in `Content Items` page and selected `Export to Deployment Target` in Action button options.
- I chose File download option and [Exportcontenttodeploymentplantarget.zip](https://github.com/OrchardCMS/OrchardCore/files/15159567/Exportcontenttodeploymentplantarget.zip) file.
- I deleted all existing content items to make sure I can import them later.
- I went to `Package Import` menu and imported `Exportcontenttodeploymentplantarget.zip` successfully. No any error messages in UI or  App_Data/Logs/orchard-log-xxx.log.
- Then I went to `Content Items` page and expected to see a content item that I had just imported but there was no any content item listed. 

Not sure what I have missed.
I hope if you can help me finding an issue since I have many content items to export and import. 
FYI, I still use Orchard Core version 1.6.0 on a production environment. 

Thanks.

**Comment by aaronamm** [comment]
Updated to 1.7.2 does not fix an issue.
My environment can't be updated to .NET 8 right now so I can't use OC 1.8.3.

**Comment by Piedone** [comment]
Importing your package worked for me with a site set up with the built-in Blog recipe with the latest `main` source. That includes a Page content type with Autoroute Part and Flow Part. I think you also need an alias-providing part, like Alias Part or Autoroute Part, for it to work.

**Reply by aaronamm** [reply]
@Piedone thank you so much. 
My content type contains only TitlePart and does not work even changing to 1.8.3 version.

I think your suggest will work.
BTW, should we improve export/import system to not require a specific part attached to a content item.

**Reply by aaronamm** [reply]
@Piedone adding Autoroute part to my Page type does not work!!
Is it possible for you to to create a simple type as my Page type?
Type: Page
Parts: [TitlePart, AutoroutePart]
![image](https://github.com/OrchardCMS/OrchardCore/assets/2938310/ee620c38-d3e4-4f00-b061-43d8e0348418)

Then, create a new Page content item, export/import and verify if you have an imported item in Content Items page.

FYI, I started a project with a blank site.

Thank you so much.

**Reply by Piedone** [reply]
I suggest you start with the Blog's Page type, and remove things what you don't need from it, one by one. That way, you can always go from steps that you've seen working.

**Reply by aaronamm** [reply]
@Piedone thank you for your suggest. I will setup with Blog recipe and report the result.

**Reply by aaronamm** [reply]
@Piedone this is very strange. I exported a BlogPost from a new setup OC 1.7.2 with Blog recipe and it didn't work. 
I tried both import with .zip file and JSON and both methods didn't work.
Here are JSON of exported content item.

```json
{
  "name": "",
  "displayName": "",
  "description": "",
  "author": "",
  "website": "",
  "version": "",
  "issetuprecipe": false,
  "categories": [],
  "tags": [],
  "steps": [
    {
      "name": "Content",
      "data": [
        {
          "ContentItemId": "4p0v5csdr7bsawvfkr9x1e26np",
          "ContentItemVersionId": "48pkq2rq50h25r4dhzrkawfs5p",
          "ContentType": "BlogPost",
          "DisplayText": "Man must explore, and this is exploration at its greatest",
          "Latest": true,
          "Published": true,
          "ModifiedUtc": "2024-05-02T17:08:24.6528714Z",
          "PublishedUtc": "2024-05-02T17:08:24.6653839Z",
          "CreatedUtc": "2024-05-02T17:08:24.6509525Z",
          "Owner": "4wv83jne67mzf1wbry511phaxh",
          "Author": "admin",
          "TitlePart": {
            "Title": "Man must explore, and this is exploration at its greatest"
          },
          "ContainedPart": {
            "ListContentItemId": "4y5yn7b3wwfndx6kahd5qnhwnh",
            "Order": 0
          },
          "MarkdownBodyPart": {
            "Markdown": "Never in all their history have men been able truly to conceive of the world as one: a single sphere, a globe, having the qualities of a globe, a round earth in which all the directions eventually meet, in which there is no center because every point, or none, is center — an equal earth which all men occupy as equals. The airman's earth, if free men make it, will be truly round: a globe in practice, not in theory.\n\nScience cuts two ways, of course; its products can be used for both good and evil. But there's no turning back from science. The early warnings about technological dangers also come from science.\n\nWhat was most significant about the lunar voyage was not that man set foot on the Moon but that they set eye on the earth.\n\nA Chinese tale tells of some men sent to harm a young girl who, upon seeing her beauty, become her protectors rather than her violators. That's how I felt seeing the Earth for the first time. I could not help but love and cherish her.\n\nFor those who have seen the Earth from space, and for the hundreds and perhaps thousands more who will, the experience most certainly changes your perspective. The things that we share in our world are far more valuable than those which divide us."
          },
          "AutoroutePart": {
            "Path": "blog/post-1"
          },
          "BlogPost": {
            "Subtitle": {
              "Text": "Problems look mighty small from 150 miles up"
            },
            "Image": {
              "Paths": [
                "post-bg.jpg"
              ],
              "MediaTexts": [
                ""
              ],
              "Anchors": [
                {
                  "X": 0.5,
                  "Y": 0.5
                }
              ]
            },
            "Tags": {
              "TermContentItemIds": [
                "4zfdqfe31bfwe1fm70k83d9pdt",
                "4r5wf3137r83t1w9pneaqav5ne",
                "4c47vjx8p0zb50nqf435vbqgd9"
              ],
              "TaxonomyContentItemId": "4j6xfagmgedqnyxvqkbzh55vce",
              "TagNames": [
                "Earth",
                "Exploration",
                "Space"
              ]
            },
            "Category": {
              "TermContentItemIds": [
                "4v71m96xevq1qy3wjmy8yfsnxp"
              ],
              "TaxonomyContentItemId": "4t6ft390rr9vqt7m61y1bf56x6"
            }
          }
        }
      ]
    }
  ]
}
```

**Reply by aaronamm** [reply]
@Piedone It didn't work on OC 1.8.3 as well. Tested.

**Reply by Piedone** [reply, marked as answer]
What exactly do you mean by it not working? I just tried with the least source, and it works, see below. So, if anything was broken, we fixed it in the meantime.

https://github.com/OrchardCMS/OrchardCore/assets/1976647/77980f67-949b-4701-8843-1fc9edfe9e90

**Comment by aaronamm** [comment]
@Piedone thank you so much for the screenshot video. It is very helpful.
Here is what I found and can tell us why this export/import tool does not work for me.

## Changes of blog post across ContentItemIndex and Document tables after importing an exported file.
- c as ContentItemIndex table
- d as Document table

| # | Action                                                                                                                           | c.Id | c.DocumentId | c.Published | c.ContentItemId            | c.ContentType | c.DisplayText                                             | d.Id | d.Content ($.ContentItemId) | d.Content ($.DisplayText)                                 |
| - | -------------------------------------------------------------------------------------------------------------------------------- | ---- | ------------ | ----------- | -------------------------- | ------------- | --------------------------------------------------------- | ---- | --------------------------- | --------------------------------------------------------- |
| 1 | Setup a new website with auto created blog post (Blog recipe)                                                                    | 5    | 10           | 1           | 46kgd55ymqfzfsw62gzksycsw6 | BlogPost      | Man must explore, and this is exploration at its greatest | 10   | 46kgd55ymqfzfsw62gzksycsw6  | Man must explore, and this is exploration at its greatest |
| 2 | Enable Export Content To Deployment Target and export an existing blog post                                                      | 5    | 10           | 1           | 46kgd55ymqfzfsw62gzksycsw6 | BlogPost      | Man must explore, and this is exploration at its greatest | 10   | 46kgd55ymqfzfsw62gzksycsw6  | Man must explore, and this is exploration at its greatest |
| 3 | Delete an exisiting blog post                                                                                                    | 14   | 10           | 0           | 46kgd55ymqfzfsw62gzksycsw6 | BlogPost      | Man must explore, and this is exploration at its greatest | 10   | 46kgd55ymqfzfsw62gzksycsw6  | Man must explore, and this is exploration at its greatest |
| 4 | Import an exported zip file without editing any JSON content by using Import Deployment Package tool                             | 14   | 10           | 0           | 46kgd55ymqfzfsw62gzksycsw6 | BlogPost      | Man must explore, and this is exploration at its greatest | 10   | 46kgd55ymqfzfsw62gzksycsw6  | Man must explore, and this is exploration at its greatest |
| 5 | Extract exported zip file, edit only DisplayText and import with "JSON Import" tool                                              | 16   | 10           | 1           | 46kgd55ymqfzfsw62gzksycsw6 | BlogPost      | New display text                                          | 10   | 46kgd55ymqfzfsw62gzksycsw6  | New display text                                          |
| 6 | Extract exported zip file, edit only ContentItemId (26 characters length), AutoroutePart.Path and import with "JSON Import" tool | 18   | 20           | 1           | 55kgd55ymqfzfsw62gzksycsw6 | BlogPost      | Man must explore, and this is exploration at its greatest | 20   | 55kgd55ymqfzfsw62gzksycsw6  | Man must explore, and this is exploration at its greatest |


From the action/result table above, here are what I found:
- In `action #3`, delete a content item does not remove a content item from a database. It is only unpublish a content item and insert a new ContentItemIndex record with unpublished state.
- In `action #4`, importing a content item without changing **DisplayText value** does not make a tool to import a content item. 
I think an import tool does not find any change in **ContentItemIndex table** of an existing content. This is because we imported a content that contains an existing ContentItemId. **Here is the issue I that found**
- In `action #5`, importing a content item with changing **DisplayText value** cause an existing content item get published with a new **ContentItemIndex record**. (This is the same as what you did in your screenshot video.)
- In `action #6`, importing a content item with changing ContentItemId and AutoroutePath creates a new Document and ContentItemIndex record. This means we have a new content item.

For those who import an exported file into another websites won't find this issue. 
However, I export and import a content item to the same website and that's why this does not work for me.
I did this because I want to make sure import/export tool works before I apply it to my production environment. 

Since deleting an content item only unpublish it, this can make a strange behavior for import/export tool.
I think for those who really want deleting an content item to really remove it from a database,  they can create a custom ContentHandler and register it in Startup.cs.
For example:
```cs
    public class BlogPostContentHandler : ContentHandlerBase
    {
        private readonly ISession _session;
        public BlogPostContentHandler(ISession session) => _session = session;

        public override async Task RemovedAsync(RemoveContentContext context)
        {
            if (context.ContentItem.ContentType != "BlogPost")
            {
                return;
            }

            _session.Delete(context.ContentItem);
            await _session.SaveChangesAsync();
        }
    }
```
In `Startup.cs`.
```cs
 services.AddScoped<IContentHandler, BlogPostContentHandler>();
```

**Reply by Piedone** [reply]
I'm getting a bit overloaded here. Could you please formulate what you need help with and couldn't figure out from the OC source as as a single question in a paragraph?

**Reply by aaronamm** [reply]
@Piedone thank you so much for following up.
I think I now understand how OC import/export works.
My problem is starting from OC does not actually delete a content item, just only unpublish it.
Therefore, if we export a content item and delete that content item.
Then later, try to import it without editing any value like `DisplayText`, that content item won't be imported.

This won't be an issue if OC actually delete a content item from a database. 

I will mark my above comment as an answer because we don't have to do anything further unless we want to change the current behavior ,e.g. allow importing a content to an OC website which have the same content item but set as unpublish.

**Reply by Piedone** [reply]
Yes, OC does only soft deletes. We may have destroy later: https://github.com/OrchardCMS/OrchardCore/issues/6056.

Since I've been assisting you here for days, be so kind and select my message as answer, not your own :).

**Reply by aaronamm** [reply]
@Piedone  thank you so much. I have updated your screenshot video as answer.

**Reply by Piedone** [reply]
Thanks!
